src/dask_awkward/lib/io/columnar.py

```diff
@@ -86,55 +87,69 @@ def project(
-        path_to_child_paths = state["path_to_child_paths"]
-        # Inverse parentage
-        form_key_to_child_keys: dict[str, list[str]] = {}
-        for child_key, parent_key in form_key_to_parent_key.items():
-            form_key_to_child_keys.setdefault(parent_key, []).append(child_key)
 
         # Require the data of metadata buffers above shape-only requests
         data_buffers = {
             *report.data_touched,
             *buffer_keys_required_to_compute_shapes(
                 parse_buffer_key,
                 report.shape_touched,
-                form_key_to_parent_key,
+                form_key_to_parent_form_key,
                 form_key_to_buffer_keys,
             ),
         }
 
         # We can't read buffers directly, but if we encounter a metadata
         # buffer, then we should be able to pick any child.
         paths = set()
-        wildcard_paths = set()
+        wildcard_form_key = set()
         for buffer_key in data_buffers:
             form_key, attribute = parse_buffer_key(buffer_key)
-            path = form_key_to_path[form_key]
             if attribute in METADATA_ATTRIBUTES:
-                wildcard_paths.add(path)
+                wildcard_form_key.add(form_key)
             else:
-                paths.add(path)
+                paths.add(form_key_to_path[form_key])
 
         # Select the most appropriate column for each wildcard
-        for path in wildcard_paths:
-            child_paths = path_to_child_paths[path]
-
-            # This is a leaf! Therefore, we read this column
-            if not child_paths:
-                paths.add(path)
+        for form_key in wildcard_form_key:
+            # Find (DFS) any non-empty record form in any child
+            recursive_child_forms = (
+                form_key_to_form[k]
+                for k in walk_graph_depth_first(form_key, form_key_to_child_form_keys)
+            )
+            record_form_keys_with_contents = (
+                f.form_key
+                for f in recursive_child_forms
+                if isinstance(f, ak.forms.RecordForm) and f.contents
+            )
+            # Now find the deepest of such records
+            try:
+                last_record_form_key = next(record_form_keys_with_contents)
+            except StopIteration:
+                # This is a leaf! Therefore, we read this column
+                paths.add(form_key_to_path[form_key])
+                continue
+            else:
+                # Ensure we get the "actual" last form key
+                for last_record_form_key in record_form_keys_with_contents:
+                    ...
+
+            # First see if any child is already included
+            for any_child_form_key in form_key_to_child_form_keys[last_record_form_key]:
+                any_child_path = form_key_to_path[any_child_form_key]
+                if any_child_path in paths:
+                    break
+            # Otherwise, add the last child
```

Comment:
@douglasdavis what do you think of this optimisation?

This is the simplest "find a leaf" that I could come up with. I originally played around with computing a heuristic for the mean buffer size of the path to the leaf, but I think it's probably not worth the complexity; any index nodes will immediately throw off the estimate. It might be reasonable to assume that most input sources won't have any index nodes, but it is a lot more work.

Reply:
OK this seems reasonable to me; maybe one day we can revisit trying to find an int leaf over e.g. a string leaf, but this should be on par with what we were already doing, so sounds good to me!